subsys/settings/src/settings_zms.c

```diff
@@ -96,117 +133,262 @@ static int settings_zms_unlink_ll_node(struct settings_zms *cf, uint32_t name_ha
 		 */
 		cf->last_hash_id = settings_element.previous_hash;
 	}
-	/* update the previous element */
-	if (settings_element.previous_hash) {
-		rc = zms_read(&cf->cf_zms, settings_element.previous_hash, &settings_update_element,
-			      sizeof(struct settings_hash_linked_list));
-		if (rc < 0) {
-			return rc;
-		}
-		if (!settings_element.next_hash) {
-			/* we are deleting the last element of the linked list,
-			 * let's update the second_to_last_hash_id
-			 */
-			cf->second_to_last_hash_id = settings_update_element.previous_hash;
-		}
-		settings_update_element.next_hash = settings_element.next_hash;
-		rc = zms_write(&cf->cf_zms, settings_element.previous_hash,
-			       &settings_update_element, sizeof(struct settings_hash_linked_list));
-		if (rc < 0) {
-			return rc;
-		}
-	}
 
-	return rc;
+	return 0;
 }
+#endif /* CONFIG_SETTINGS_ZMS_NO_LL_DELETE */
 
 static int settings_zms_delete(struct settings_zms *cf, uint32_t name_hash)
 {
 	int rc = 0;
 
 	rc = zms_delete(&cf->cf_zms, name_hash);
 	if (rc >= 0) {
-		rc = zms_delete(&cf->cf_zms, name_hash + ZMS_DATA_ID_OFFSET);
+		rc = zms_delete(&cf->cf_zms, ZMS_DATA_ID_FROM_NAME(name_hash));
 	}
 	if (rc < 0) {
 		return rc;
 	}
 
+#ifndef CONFIG_SETTINGS_ZMS_NO_LL_DELETE
+#ifdef CONFIG_SETTINGS_ZMS_LL_CACHE
+	cf->ll_has_changed = true;
+#endif
 	rc = settings_zms_unlink_ll_node(cf, name_hash);
-	if (rc < 0) {
-		return rc;
+#endif /* CONFIG_SETTINGS_ZMS_NO_LL_DELETE */
+
+	return rc;
+}
+
+#ifdef CONFIG_SETTINGS_ZMS_LOAD_SUBTREE_PATH
+/* Loads first the key which is defined by the name found in "subtree" root.
+ * If the key is not found or further keys under the same subtree are needed
+ * by the caller, returns 0.
+ */
+static int settings_zms_load_subtree(struct settings_store *cs, const struct settings_load_arg *arg)
+{
+	struct settings_zms *cf = CONTAINER_OF(cs, struct settings_zms, cf_store);
+	struct settings_zms_read_fn_arg read_fn_arg;
+	char name[SETTINGS_FULL_NAME_LEN];
+	ssize_t rc1;
+	ssize_t rc2;
+	uint32_t name_hash;
+
+	name_hash = sys_hash32(arg->subtree, strnlen(arg->subtree, SETTINGS_FULL_NAME_LEN)) &
+		    ZMS_HASH_MASK;
+	for (int i = 0; i <= cf->hash_collision_num; i++) {
+		name_hash = ZMS_UPDATE_COLLISION_NUM(name_hash, i);
+		/* Get the name entry from ZMS */
+		rc1 = zms_read(&cf->cf_zms, ZMS_NAME_ID_FROM_HASH(name_hash), &name,
+			       sizeof(name) - 1);
+		/* get the length of data and verify that it exists */
+		rc2 = zms_get_data_length(&cf->cf_zms, ZMS_DATA_ID_FROM_HASH(name_hash));
+		if ((rc1 <= 0) || (rc2 <= 0)) {
+			/* Name or data doesn't exist */
+			continue;
+		}
+		/* Found a name, this might not include a trailing \0 */
+		name[rc1] = '\0';
+		if (strcmp(arg->subtree, name)) {
+			/* Names are not equal let's continue to the next collision hash
+			 * if it exists.
+			 */
+			continue;
+		}
+		/* At this steps the names are equal, let's set the handler */
+		read_fn_arg.fs = &cf->cf_zms;
+		read_fn_arg.id = ZMS_DATA_ID_FROM_HASH(name_hash);
```

Comment:
This is not exactly an ID right? Since ZMS_DATA_ID_FROM_HASH is simply setting bit 31 and adding offset to HASH. Only when you re-hash this "id", you would get the "actual ID", correct?

Or in this logic you are simply using 'hashes' as data IDs?


So if name hash is:
0x01234567 (let's say corresponding to name ID 0x00000001), then value id becomes:
0x41234567 (which could correspond to 0x40000001, but now, the system stores only hashed ID on the NVM and not the real IDs that you would expect)

Is it correct?

Reply:
Never mind. I think the confusion comes because names are written using name IDs.